### PR TITLE
Structured reference condition

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

When developing references, we postponed certain features associated with it in order to ship the change ahead of a new cycle.

One of the things that we postponed, due to the fact that users could make use of existing conditions functionality, was the addition of a structured reference condition.

## What it is not done yet?

- [ ] Candidate interface
- [ ] Change (edit) offer

## Screenshots

Screenshot 2023-04-14 at 11 24 37
Screenshot 2023-04-14 at 11 25 04
Screenshot 2023-04-14 at 11 35 34

### CODE REVIEW

**Comment by tomas-stefano** [comment]
[Create a structured reference condition for references (Manage users)](https://trello.com/c/zNDHkb5N/1347-create-a-structured-reference-condition-for-references-manage-users)

File: spec/forms/provider_interface/offer_wizard_spec.rb [hunk 1]
```diff
@@ -635,4 +635,142 @@ def further_conditions_array
...
+      context 'when require references does not have any value' do
+        it 'returns checked' do
+          expect(wizard.require_references).to be(1)
+        end
+      end
+
+      context 'when require references is checked' do
+        it 'returns checked' do
+          wizard.require_references = '1'
+          expect(wizard.require_references).to be(1)
```
Comment (carlosmartinez) [inline spec/forms/provider_interface/offer_wizard_spec.rb:654]: this reads a little bit like
```
a = 1
expect(a).to be(1)
```
is there more to it than that?

File: app/models/reference_condition.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  detail :description
+
+  def initialize(attrs = {})
+    attrs ||= {}
+    super({ status: :pending }.merge(attrs))
+  end
+
+  def structured_condition?
+    true
+  end
```
Comment (carlosmartinez) [inline app/models/reference_condition.rb:12]: is this a useful thing?

File: app/components/provider_interface/conditions_list_component.html.erb [hunk 3]
```diff
@@ -4,6 +4,9 @@
       <div class="govuk-summary-list__row<%= ' govuk-summary-list__row--no-border' if index == conditions.length - 1 %>">
         <dt class="govuk-summary-list__key">
           <%= condition.text %>
+          <% if condition.respond_to?(:structured_condition?) && condition.description.present? %>
```
Comment (carlosmartinez) [inline app/components/provider_interface/conditions_list_component.html.erb:7]: could we not just do

<% if condition.description %>

**Review by stevehook — COMMENTED** [review]
This all worked for me locally except that when I re-edited a reference condition that I'd entered the description part wasn't loaded back into the edit form.

I left a couple of minor questions inline but it's all looking good.

(I also found a bug in the SKE condition code that I added).

File: app/forms/provider_interface/offer_wizard.rb [hunk 4]
```diff
@@ -46,12 +48,41 @@ def self.build_from_application_choice(state_store, application_choice, options
...
     end
 
+    def require_references
+      return REQUIRE_REFERENCES_CHECKED_BY_DEFAULT if @require_references.nil? && FeatureFlag.active?(:structured_reference_condition)
+
+      @require_references.to_i
+    end
+
+    def require_references?
+      @require_references.present? && require_references.nonzero?.present?
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/forms/provider_interface/offer_wizard.rb:58]: This is the creation only (only Redis before the DB) so it is not stored yet. Wizards 😢

File: app/models/reference_condition.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/models/reference_condition.rb:12]: I will remove this method and leave the "problem" to the next structured condition. Not a tech debt though.

**Review by stevehook — APPROVED** [review]

**Pull request merged**